research-software-directory/RSD-as-a-service
# 234 login options

# Flexible login options

Closes #234 

Changes proposed in this pull request:
* RSD instance can be defined through env. variable RSD_AUTH_PROVIDERS to offer the login with SURFconext, Helmholtz AAI or both
* When one provider is defined the Sign in button redirects directly (as in the past)
* When more than one provider is defined, the modal with the list of the providers is shown. The order of providers in the list is identical to the order of values in the env variable. The values of the providers should be semicolon (;) separated.
* When RSD_AUTH_PROVIDERS is missing, the default provider is SURFconext
* The default values are specified in the api functions for each provider in case that env variable is missing (please check if defined defaults are correct). The name of the provider, used in the button, is also specified there.

How to test:
* Extend you .env file with RSD_AUTH_PROVIDERS, see env.example file
* Specify one or both providers (not case sensitive, semicolon separated)
   *  RSD_AUTH_PROVIDERS=surfconext;helmholtzaai
   *  RSD_AUTH_PROVIDERS=surfconext
   *  RSD_AUTH_PROVIDERS=helmholtzaai
* `docker-compose build`: to build front-end code
* `docker-compose up`: to start
* Try without, with one and both providers in the env file
* Testing login with both providers. I was able to login with SURFconext but not with Helmholtz AAI (using ORCID) on localhost. The Java code logs error about the organisation

## Suggestion for further improvements

- Use well-know endpoint url to obtain the token url instead of AUTH_HELMHOLTZAAI_TOKEN_URL, AUTH_SURFCONEXT_TOKEN_URL. I would advice this improvement in the auth module (Java)
- ~~Investigate why Helmholtz AAI fails with ORCID. I was able to login on remote version with ORCID, however the localhost does not work~~  **Resolved by using flag HELMHOLTZAAI_ALLOW_EXTERNAL_USERS**

## Example multiple providers

![image](https://user-images.githubusercontent.com/9204081/166294808-aef84984-8e5d-4c74-bff0-bf665993a6a5.png)

PR Checklist:
